TowerScheduler/scheduler.py

```diff
@@ -0,0 +1,442 @@
+    '''
+
+    singleton = None
+
+    def __init__(self):
+        self.config = Configuration.get_singleton(Path('schedulerConfig.ini'))
+        self.logger = get_logger("Iterate State", self.config.log_level)
+
+    @classmethod
+    def get_singleton(cls):
+        if Iterate.singleton is None:
+            Iterate.singleton = Iterate()
+        return Iterate.singleton
+
+    def process(self, state_machine):
+        self.logger.info("Running Iterate process func")
+        state_machine.ens_index += 1
+
+    def update(self, state_machine):
+        self.logger.info("Running Iterate update func")
+        state_machine.curr_state = CheckTime.get_singleton()
+
+
+class PerformEnsemble(State):
+    '''
+    PerformEnsemble runs the function associated with the current ensemble
+    and then passes to Iterate
+    '''
+
+    singleton = None
+
+    def __init__(self):
+        self.config = Configuration.get_singleton(Path('schedulerConfig.ini'))
+        self.logger = get_logger("Perform Ensemble State",
+                                self.config.log_level)
+
+    @classmethod
+    def get_singleton(cls):
+        if PerformEnsemble.singleton is None:
+            PerformEnsemble.singleton = PerformEnsemble()
+        return PerformEnsemble.singleton
+
+    def process(self, state_machine):
+        self.logger.info("Running PERFORM process func")
+
+        curr_ens = state_machine.ens_list[state_machine.ens_index]
+
+        curr_ens.perform_ensemble_function()
+
+        self.logger.info("Done performing %s", curr_ens.title)
+
+        now = dt.datetime.now()
+        self.logger.info("Time is now %02i:%02i:%02i",
+                now.hour, now.minute, now.second)
+
+    def update(self, state_machine):
+        self.logger.info("Running PERFORM update func")
+        state_machine.curr_state = Iterate.get_singleton()
+
+
+class Sleep(State):
+    '''
+    Sleep calculates how much time is available to sleep once
+    the need for sleep is confirmed. A hardware sleep will
+    make the machine reset and wake up in WakeUp. A software
+    sleep leads to CheckTime.
+    '''
+
+    singleton = None
+
+    def __init__(self):
+        self.nearest_ens_time: dt.time = None
+        self.config = Configuration.get_singleton(Path('schedulerConfig.ini'))
+        self.logger = get_logger("Sleep State", self.config.log_level)
+
+    @classmethod
+    def get_singleton(cls):
+        if Sleep.singleton is None:
+            Sleep.singleton = Sleep()
+        return Sleep.singleton
+
+    def process(self, state_machine):
+        self.logger.info("Running Sleep process func")
+
+        # write curr index to ens file before calcs
+        with open("active_ensembles.json", "w", encoding="utf-8") as f_out:
+            json_file = {
+                "ensemble_list": Ensemble.list_to_json(state_machine.ens_list),
+                "next_ensemble": state_machine.ens_index
+            }
+            json.dump(json_file, f_out, indent=4)
```

Comment:
To clarify, this is dangerous because we could crash in the middle of writing the ensemble list, thus losing the list as a whole since we don't have any sort of backup copy, correct? If my understanding is correct, this would be solved by separating the ensemble list from the next ensemble index and saving them in two separate files, as you suggested doing in another comment. I wanted to confirm that this would still be a safe location to write the next ensemble index, as long as we aren't trying to write the list as well.